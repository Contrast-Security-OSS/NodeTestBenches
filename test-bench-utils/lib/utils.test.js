'use strict';

const utils = require('./utils');

jest.mock('./frameworks', () => ({
  framework: {
    query: { method: 'get', key: 'query' },
    params: { method: 'get', key: 'params', param: ':input' },
    headers: { method: 'get', key: 'headers' },
    body: { method: 'post', key: 'body' },
    cookies: { method: 'post', key: 'cookies' },
    input: { method: 'get' }
  }
}));

jest.mock('./routes', () => ({
  rule: {
    base: '/rule',
    name: 'OWASP Rule 123',
    link: 'https://example.com',
    products: ['Assess', 'Protect'],
    inputs: ['query', 'params', 'body'],
    params: ['input', 'second_input'],
    additionalData: ['foo', 'bar'],
    sinks: {
      rule: jest.fn()
    }
  },
  rule2: {
    base: '/rule2',
    name: 'OWASP Rule 345',
    link: 'https://example.com',
    products: ['Assess', 'Protect'],
    inputs: ['query', 'params', 'body'],
    params: ['input', 'second_input'],
    additionalData: ['foo', 'bar'],
    sinks: {
      rule2: {
        noop: jest.fn(),
        unsafe: jest.fn(),
        safe: jest.fn(),
        safe2: jest.fn()
      }
    }
  }
}));

describe('getSinkData', function() {
  it('returns the expected array of SinkData objects for each input', function() {
    const result = utils.getSinkData('rule', 'framework');
    expect(result).toMatchSnapshot();

    const mockSink = require('./routes').rule.sinks.rule;

    result[0].sinks.safe({ input: 'foo' });
    expect(mockSink).toHaveBeenCalledWith({ input: 'foo' }, { safe: true });

    result[1].sinks.unsafe({ input: 'bar' });
    expect(mockSink).toHaveBeenCalledWith({ input: 'bar' });

    result[2].sinks.noop({ input: 'baz' });
    expect(mockSink).toHaveBeenCalledWith({ input: 'baz' }, { noop: true });
  });

  it('returns the expected array of SinkData objects for each sink', function() {
    const result = utils.getSinkData('rule2', 'framework');
    expect(result).toMatchSnapshot();

    const mockSinks = require('./routes').rule2.sinks.rule2;

    result[0].sinks.safe({ input: 'foo' });
    expect(mockSinks.safe).toHaveBeenCalledWith({ input: 'foo' });

    result[1].sinks.safe2({ input: 'fooo' });
    expect(mockSinks.safe2).toHaveBeenCalledWith({ input: 'fooo' });

    result[2].sinks.unsafe({ input: 'bar' });
    expect(mockSinks.unsafe).toHaveBeenCalledWith({ input: 'bar' });

    result[0].sinks.noop({ input: 'baz' });
    expect(mockSinks.noop).toHaveBeenCalledWith({ input: 'baz' });
  });
});

describe('groupSinkData', function() {
  it('returns the expected grouped object', function() {
    const sinkData = utils.getSinkData('rule', 'framework');
    const result = utils.groupSinkData(sinkData);
    expect(result).toMatchSnapshot();
  });
});

describe('getRouteMeta', function() {
  it('returns the metadata for a given route', function() {
    const result = utils.getRouteMeta('rule');
    expect(result).toMatchObject({
      base: '/rule',
      name: 'OWASP Rule 123',
      link: 'https://example.com',
      products: ['Assess', 'Protect'],
      inputs: ['query', 'params', 'body'],
      params: ['input', 'second_input'],
      additionalData: ['foo', 'bar'],
      sinks: {
        rule: expect.any(Function)
      }
    });
  });
});

describe('getInput', function() {
  const key = 'body';
  const params = ['input', 'arg'];
  let req;

  beforeEach(function() {
    req = {
      body: {
        input: 'foo',
        arg: 'bar',
        something: 'else'
      }
    };
  });

  it('returns the appropriate object keyed by expected parameters', function() {
    const result = utils.getInput(req, key, params);
    expect(result).toEqual({
      input: 'foo',
      arg: 'bar'
    });
  });

  it('returns parameters from the locals object when passed with input params', function() {
    const locals = { input: 'FOO', arg: 'BAZ', attackValues: 'bad string!' };
    const result = utils.getInput(req, key, params, { locals });
    expect(result).toEqual({
      input: 'FOO',
      arg: 'BAZ'
    });
  });

  it('returns hard-coded values when provided', function() {
    const result = utils.getInput(req, key, params, { noop: true });
    expect(result).toEqual({
      input: 'noop',
      arg: 'noop'
    });
  });
});
