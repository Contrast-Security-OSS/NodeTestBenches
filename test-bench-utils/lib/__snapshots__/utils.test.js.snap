// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSinkData returns the expected array of SinkData objects for each input 1`] = `
Array [
  Object {
    "input": "query",
    "key": "query",
    "method": "get",
    "name": "rule",
    "params": Array [
      "input",
      "second_input",
    ],
    "sinks": Object {
      "noop": [Function],
      "safe": [Function],
      "unsafe": [Function],
    },
    "uri": "/query/rule",
    "url": "/rule/query/rule",
    "urlWithoutParams": "/rule/query/rule",
  },
  Object {
    "input": "params",
    "key": "params",
    "method": "get",
    "name": "rule",
    "params": Array [
      "input",
      "second_input",
    ],
    "sinks": Object {
      "noop": [Function],
      "safe": [Function],
      "unsafe": [Function],
    },
    "uri": "/params/rule/:input",
    "url": "/rule/params/rule/:input",
    "urlWithoutParams": "/rule/params/rule",
  },
  Object {
    "input": "body",
    "key": "body",
    "method": "post",
    "name": "rule",
    "params": Array [
      "input",
      "second_input",
    ],
    "sinks": Object {
      "noop": [Function],
      "safe": [Function],
      "unsafe": [Function],
    },
    "uri": "/body/rule",
    "url": "/rule/body/rule",
    "urlWithoutParams": "/rule/body/rule",
  },
]
`;

exports[`getSinkData returns the expected array of SinkData objects for each sink 1`] = `
Array [
  Object {
    "input": "query",
    "key": "query",
    "method": "get",
    "name": "rule2",
    "params": Array [
      "input",
      "second_input",
    ],
    "sinks": Object {
      "noop": [MockFunction],
      "safe": [MockFunction],
      "safe2": [MockFunction],
      "unsafe": [MockFunction],
    },
    "uri": "/query/rule2",
    "url": "/rule2/query/rule2",
    "urlWithoutParams": "/rule2/query/rule2",
  },
  Object {
    "input": "params",
    "key": "params",
    "method": "get",
    "name": "rule2",
    "params": Array [
      "input",
      "second_input",
    ],
    "sinks": Object {
      "noop": [MockFunction],
      "safe": [MockFunction],
      "safe2": [MockFunction],
      "unsafe": [MockFunction],
    },
    "uri": "/params/rule2/:input",
    "url": "/rule2/params/rule2/:input",
    "urlWithoutParams": "/rule2/params/rule2",
  },
  Object {
    "input": "body",
    "key": "body",
    "method": "post",
    "name": "rule2",
    "params": Array [
      "input",
      "second_input",
    ],
    "sinks": Object {
      "noop": [MockFunction],
      "safe": [MockFunction],
      "safe2": [MockFunction],
      "unsafe": [MockFunction],
    },
    "uri": "/body/rule2",
    "url": "/rule2/body/rule2",
    "urlWithoutParams": "/rule2/body/rule2",
  },
]
`;

exports[`groupSinkData returns the expected grouped object 1`] = `
Object {
  "body": Array [
    Object {
      "input": "body",
      "key": "body",
      "method": "post",
      "name": "rule",
      "params": Array [
        "input",
        "second_input",
      ],
      "sinks": Object {
        "noop": [Function],
        "safe": [Function],
        "unsafe": [Function],
      },
      "uri": "/body/rule",
      "url": "/rule/body/rule",
      "urlWithoutParams": "/rule/body/rule",
    },
  ],
  "params": Array [
    Object {
      "input": "params",
      "key": "params",
      "method": "get",
      "name": "rule",
      "params": Array [
        "input",
        "second_input",
      ],
      "sinks": Object {
        "noop": [Function],
        "safe": [Function],
        "unsafe": [Function],
      },
      "uri": "/params/rule/:input",
      "url": "/rule/params/rule/:input",
      "urlWithoutParams": "/rule/params/rule",
    },
  ],
  "query": Array [
    Object {
      "input": "query",
      "key": "query",
      "method": "get",
      "name": "rule",
      "params": Array [
        "input",
        "second_input",
      ],
      "sinks": Object {
        "noop": [Function],
        "safe": [Function],
        "unsafe": [Function],
      },
      "uri": "/query/rule",
      "url": "/rule/query/rule",
      "urlWithoutParams": "/rule/query/rule",
    },
  ],
}
`;
